README.md
# 835Parser

Parse 835 files to quickly pull file content for cross referencing. Program outputs a file containing rows for each CLP and PLB segment found within the parsed files.

Rows of data contain the following columns:

- FILENAME
- TRN02
- TRN03
- PAYER
- PAYEE
- NPI
- CLAIM (CLP 01)
- CLP02
- PLB_DATA (the whole PLB segment's data)

## Installation

For stand alone usage on Windows, download and unzip the 835FileParser.zip file and run as desired. Otherwise within a python installation run Strip835UI.py.

## Usage

### 835FileParser.exe

For stand alone usage on Windows, download and unzip the 835FileParser.zip file and run as desired.

To Run:

1. Specify the Open location containing files to parse
2. Specify the desired results file output location
3. Optionally, change the default ".835" file pattern to a desired file pattern
    - Files will qualify for parsing if they contain or end with the indicated file pattern

From the Settings menu:

- Toggle whether subsequent program runs append the results to the initial run's results.
- Toggle whether the program should search in subdirectories of the specified open location, for files matching the file pattern

![Screenshot 1](https://github.com/danielslabo/835Parser/blob/main/img/screenshot1.jpg?raw=true)

### Strip835GUI.py

Run as a direct python program via shell. Offers a GUI for a user to specify the source directory and also the output dircetory. Allows a user to specify the file pattern to use as well. Uses associated ico and png files.

### Strip835Files.py

Run as a direct python program via shell. Assumes the 835 files are in the same directory the program is located and ran from. Would need to modify the file pattern in the program here diretly. Outputs to same directory. This
version is out of date and no longer updated.

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## License

### [APACHE 2.0](https://choosealicense.com/licenses/apache-2.0/)

=======


changelog.txt
# Changelog
All notable changes to this project will be documented in this file.
For most current status see https://github.com/danielslabo/835Parser

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- Enh: Strip newlines when processing a file to handle newline inconsistencies
- Arch: Incorporate using a Queue, potentially for processing thread worker's outputs
- Error: Handle I/O, import and other errors because currently not handling any
- Arch: Implement a way to kill a thread without using a daemon thread and killing the app's process
- Arch: use after() method for updating progress bar
- Enh: Account for changes to TRN and N1* segments if more than one in a file. Unconfirmed if necessary

## [3.2.2] - 2021-10-02
### Fixed
- Made various small fixes per pep8 style guide

### Added
- Function comments using docstrings

## [3.2.1] - 2021-08-16
### Fixed
- Issue 1: Spec file corrected filename from 835FilerParser to 835FileParser
- Issue 2: Progress bar resets after a run
- Issue 4: Readme updated with information on settings
- Issue 5: Changelog window uses same icon as main app
- Issue 6: Changelog content updated

## [3.2.0] - 2021-08-15
### Added
- Exe version of program

## [3.1.1] - 2021-08-15
### Added
- Setting to enable searching subdirectories in specified open location

### Changed
- Results file prefix changed to Parsed835Results
- Decomposed widget initializations into separate functions
- Output shows full file path for results and not just filename
- Additional init variables made private
- File menu and Settings Menu are now scoped as class variables
- Help text indicates shorthand for which fields are parsed
- Progress bar no longer is started and stopped for step intervals to avoid looping progress when running long.

### Fixed
- When program searches subdirectories, it will find the reltive path to files in subdirectories
- When program runs more than once, the results file is actually created. This was a problem with
  the outfile name and path previously not being updated.
- Menu options for File>Open, File>Save, and the Settings menu items now disable while program is running.
- Append results settings option now works

## [3.1.0] - 2021-08-15
### Added
- Menu bar added with submenus of File, Settings, and Help
- Settings option added to toggle appending subsequent runs' results to initial file
- Help menu opens Help, and Changelog files

### Changed
- Made thread for processing 835 files a daemon thread so will stop when main app closes.
- Some variables were made private.
- Changed icon to one created by a free vector image.

## [3.0.0] - 2021-08-14
### Added
- Thread processing so GUI doesn't freeze while running
- Progress bar implemented

### Changed
- Decomposed parsing 835s into several methods.
- Fixed typos and reworded comments.

### Removed
- Comment section on pyinstaller steps.

### Fixed
- Program can run more than once without having to re-open.

## [2.0.0] - YYYY-MM-DD
### Added
- User can specify file pattern
- Parse PLB segment data into last column of output

### Changed
- Can handle 835s where all data is in one line of the file
- Output message box is no longer editable
- CSV writer uses a QUOTE_MINIMAL setting

### Removed
- Logic to refresh the app's fields and data should an unimplemented refresh button be clicked
- Onefile executable version of program as it was out of date

### Fixed
- Processing output's numerous extra new lines and first character missing
- Comment typos and unclear function names

## [1.0.0] - YYYY-MM-DD
### Added
- Initial Release
- Program uses a GUI for input options

### Deprecated
- Non-Gui version of program

## Explanation of Sections:
- Added for new features.
- Changed for changes in existing functionality.
- Deprecated for soon-to-be removed features.
- Removed for now removed features.
- Fixed for any bug fixes.
- Security in case of vulnerabilities.

example.py
from itertools import count
import sys
import tkinter as tk
import tkinter.messagebox as tkMessageBox
import threading
import time
from random import randint
import queue

# Based on example Dialog 
# http://effbot.org/tkinterbook/tkinter-dialog-windows.htm
class InfoMessage(tk.Toplevel):
    def __init__(self, parent, info, title=None, modal=True):
        tk.Toplevel.__init__(self, parent)
        self.transient(parent)
        if title:
            self.title(title)
        self.parent = parent

        body = tk.Frame(self)
        self.initial_focus = self.body(body, info)
        body.pack(padx=5, pady=5)

        self.buttonbox()

        if modal:
            self.grab_set()

        if not self.initial_focus:
            self.initial_focus = self
        self.protocol("WM_DELETE_WINDOW", self.cancel)
        self.geometry("+%d+%d" % (parent.winfo_rootx()+50, parent.winfo_rooty()+50))
        self.initial_focus.focus_set()

        if modal:
            self.wait_window(self)  # Wait until this window is destroyed.

    def body(self, parent, info):
        label = tk.Label(parent, text=info)
        label.pack()
        return label  # Initial focus.

    def buttonbox(self):
        box = tk.Frame(self)
        w = tk.Button(box, text="OK", width=10, command=self.ok, default=tk.ACTIVE)
        w.pack(side=tk.LEFT, padx=5, pady=5)
        self.bind("<Return>", self.ok)
        box.pack()

    def ok(self, event=None):
        self.withdraw()
        self.update_idletasks()
        self.cancel()

    def cancel(self, event=None):
        # Put focus back to the parent window.
        self.parent.focus_set()
        self.destroy()


class GuiPart:
    TIME_INTERVAL = 0.1

    def __init__(self, master, queue, end_command):
        self.queue = queue
        self.master = master
        console = tk.Button(master, text='Done', command=end_command)
        console.pack(expand=True)
        self.update_gui()  # Start periodic GUI updating.

    def update_gui(self):
        try:
            self.master.update_idletasks()
            threading.Timer(self.TIME_INTERVAL, self.update_gui).start()
        except RuntimeError:  # mainloop no longer running.
            pass

    def process_incoming(self):
        """ Handle all messages currently in the queue. """
        while self.queue.qsize():
            try:
                info = self.queue.get_nowait()
                InfoMessage(self.master, info, "Status", modal=False)
            except queue.Empty:  # Shouldn't happen.
                pass


class ThreadedClient:
    """ Launch the main part of the GUI and the worker thread. periodic_call()
        and end_application() could reside in the GUI part, but putting them
        here means all the thread controls are in a single place.
    """
    def __init__(self, master):
        self.master = master
        self.count = count(start=1)
        self.queue = queue.Queue()

        # Set up the GUI part.
        self.gui = GuiPart(master, self.queue, self.end_application)

        # Set up the background processing thread.
        self.running = True
        self.thread = threading.Thread(target=self.workerthread)
        self.thread.start()

        # Start periodic checking of the queue.
        self.periodic_call(200)  # Every 200 ms.

    def periodic_call(self, delay):
        """ Every delay ms process everything new in the queue. """
        self.gui.process_incoming()
        if not self.running:
            sys.exit(1)
        self.master.after(delay, self.periodic_call, delay)

    # Runs in separate thread - NO tkinter calls allowed.
    def workerthread(self):
        while self.running:
            time.sleep(randint(1, 10))  # Time-consuming processing.
            count = next(self.count)
            info = 'Report #{} created'.format(count)
            self.queue.put(info)

    def end_application(self):
        self.running = False  # Stop queue checking.
        self.master.quit()


if __name__ == '__main__':  # Needed to support multiprocessing.
    root = tk.Tk()
    root.title('Report Generator')
    root.minsize(300, 100)
    client = ThreadedClient(root)
    root.mainloop()  # Display application window and start tkinter event loop.

depr/Strip835Files.py
# Note this non-GUI version is deprecated
import csv
import os
import re
import time

timestr = time.strftime("%Y.%m.%d-%H%M%S")
outfile_name = "Parsed835Goodies" + " " + timestr + ".csv"
outfile_headers = ['FILENAME', 'TRN02', 'TRN03',
                   'PAYER', 'PAYEE', 'NPI', 'CLAIM', 'CLP02']
file_pattern = ".835"

source_dir = os.path.dirname(os.path.abspath(__file__))
file_exists = os.path.isfile(outfile_name)


def write_to_csv(*args):
    with open(outfile_name, newline='', mode='a') as outcsv:
        csv_writer = csv.writer(
            outcsv, delimiter=',', quotechar='"', quoting=csv.QUOTE_MINIMAL)
        data = []
        for x in args:
            data.append(x)
        csv_writer.writerow(data)


# Write out Headers if output file is being newly created in current directory
with open(outfile_name, newline='', mode='a') as outcsv:
    csv_writer = csv.writer(
        outcsv, delimiter=',', quotechar='"', quoting=csv.QUOTE_MINIMAL)
    if not file_exists:
        csv_writer.writerow(outfile_headers)

# Iterate through each '835' file in current directory and parse out desired
# values. Consider if file is split across lines (standard) or all in 1 line
# (uncommon). Write desired values out as each CLP segment is found.
for root, dirs, files in os.walk(source_dir):
    for file in files:
        if file.endswith(file_pattern) or (file_pattern in file):
            print(f'Reading file: {file}')
            file_trn02 = file_trn03 = file_payer = file_payee = file_npi = ""
            num_lines_in_file = len(open(file).readlines())
            if num_lines_in_file == 1:
                for line in open(file, 'r'):
                    one_line_data = line.split(sep="~")
                    for i in range(len(one_line_data)):
                        if one_line_data[i].startswith("TRN"):
                            file_trn02 = re.sub('~', '', one_line_data[i].split('*')[2])           # TRN;02
                            file_trn03 = re.sub('~', '', one_line_data[i].split('*')[3])           # TRN;03
                        if one_line_data[i].startswith("N1*PR"):
                            file_payer = re.sub('~', '', one_line_data[i].split('*')[2]).rstrip()  # N1;02
                        if one_line_data[i].startswith("N1*PE"):
                            file_payee = re.sub('~', '', one_line_data[i].split('*')[2]).rstrip()  # N1;02
                            file_npi = re.sub('~', '', one_line_data[i].split('*')[4])             # N1;04
                        if one_line_data[i].startswith("CLP"):
                            claim = re.sub('~', '', one_line_data[i].split('*')[1])                # CLP;01
                            clp02 = re.sub('~', '', one_line_data[i].split('*')[2])                # CLP;02
                            write_to_csv(
                                file, file_trn02, file_trn03, file_payer, 
                                file_payee, file_npi, claim, clp02)
            else:
                for line in open(file, 'r'):
                    if line.startswith("TRN"):
                        file_trn02 = re.sub('~', '', line.split('*')[2])           # TRN;02
                        file_trn03 = re.sub('~', '', line.split('*')[3])           # TRN;03
                    if line.startswith("N1*PR"):
                        file_payer = re.sub('~', '', line.split('*')[2]).rstrip()  # N1;02
                    if line.startswith("N1*PE"):
                        file_payee = re.sub('~', '', line.split('*')[2]).rstrip()  # N1;02
                        file_npi = re.sub('~', '', line.split('*')[4])             # N1;04
                    if line.startswith("CLP"):
                        claim = re.sub('~', '', line.split('*')[1])                # CLP;01
                        clp02 = re.sub('~', '', line.split('*')[2])                # CLP;02
                        write_to_csv(
                            file, file_trn02, file_trn03, file_payer, 
                            file_payee, file_npi, claim, clp02)
print(f"Completed stripping 835s in: {source_dir} to {outfile_name}\n")


Strip835GUI.py
import os
import csv
import sys
import re
import threading
import time
import tkinter as tk
from tkinter import scrolledtext, ttk, messagebox
from tkinter.filedialog import askdirectory


class App(tk.Frame):
    __OUTFILE_PREFIX = "Parsed835Results"
    __OUTFILE_HEADERS = ['FILENAME', 'TRN02', 'TRN03', 'PAYER', 'PAYEE', 'NPI',
                         'CLAIM', 'CLP02', 'PLB_DATA']
    __DEFAULT_FILE_PATTERN = ".835"
    __HELPFILE = 'help.txt'
    __CHANGELOG = 'changelog.txt'
    __ICONFILE = '835_icon.ico'

    def __init__(self, master, **kw):
        super().__init__(master, **kw)
        self.master = master
        self.icon_path = self.resource_path(self.__ICONFILE)
        self.master.iconbitmap(self.icon_path)
        self.__outfile_path = ""
        self.__source_dir = ""
        self.__file_pattern = ""
        self.__append_runs = tk.BooleanVar()
        self.__append_runs.set(True)
        self.__traverse_subdir = tk.BooleanVar()
        self.__traverse_subdir.set(False)
        self.__run_counter = 1
        self.__outfile_name = self.get_new_outfile_name()
        self.__file_exists = self.check_outfile_exists(self.__outfile_name)

        self.__init_widgets_menu()
        self.__init_widgets_other()

    def __init_widgets_menu(self):
        """Sets up menu related widgets."""
        # Menu Bar
        menu_bar = tk.Menu(self.master)
        self.master.config(menu=menu_bar)

        # File
        self.file_menu = tk.Menu(menu_bar, tearoff=0)
        self.file_menu.add_command(label="Open",
                                   command=self.browse_for_open_loc)
        self.file_menu.add_command(label="Save",
                                   command=self.browse_for_save_loc)
        self.file_menu.add_separator()
        self.file_menu.add_command(label="Exit", command=self.quit)
        menu_bar.add_cascade(label='File', menu=self.file_menu)

        # Settings
        self.settings_menu = tk.Menu(menu_bar, tearoff=0)
        self.settings_menu.add_checkbutton(
            label="Append subsequent runs", onvalue=1, offvalue=0,
            variable=self.__append_runs, command=None)
        self.settings_menu.add_checkbutton(
            label="Search in subdirectories", onvalue=1, offvalue=0,
            variable=self.__traverse_subdir, command=None)
        menu_bar.add_cascade(label='Settings', menu=self.settings_menu)

        # Help
        help_menu = tk.Menu(menu_bar, tearoff=0)
        help_menu.add_command(label="Help", command=self.open_help)
        help_menu.add_command(label="Change Log", command=self.open_changelog)
        menu_bar.add_cascade(label="Help", menu=help_menu)

    def __init_widgets_other(self):
        """Sets up non-menu widgets. """
        # Frame Setup
        inputs_frame = tk.LabelFrame(root, text="1. Enter File Details: ")
        inputs_frame.grid(row=0, columnspan=10, sticky='WE', padx=5, pady=5,
                          ipadx=5, ipady=5)
        inputs_frame.columnconfigure(1, weight=1)

        output_frame = tk.Label(root)
        output_frame.grid(row=1, columnspan=10, sticky='NSEW', padx=5, pady=5,
                          ipadx=5, ipady=5)
        output_frame.columnconfigure(1, weight=1)
        output_frame.rowconfigure(0, weight=1)

        progress_frame = tk.Frame(root)
        progress_frame.grid(row=2, columnspan=10, sticky='WE', padx=5, pady=5)
        progress_frame.columnconfigure(1, weight=1)
        progress_frame.rowconfigure(1, weight=1)

        footer_frame = tk.Label(root)
        footer_frame.grid(row=3, columnspan=10, sticky='EW', padx=5, pady=1)
        footer_frame.columnconfigure(1, weight=1)
        footer_frame.rowconfigure(2, weight=1)

        # File Pattern Input
        file_pattern_lbl = tk.Label(inputs_frame, text="File Pattern", anchor='w')
        file_pattern_lbl.grid(row=0, column=0, sticky='WE', padx=5, pady=2)

        self.file_pattern_txt = tk.Entry(inputs_frame, state="normal")
        self.file_pattern_txt.grid(row=0, column=1, columnspan=7,
                                   sticky="WE", padx=5, pady=2)
        self.file_pattern_txt.insert(0, str(self.__DEFAULT_FILE_PATTERN))

        # Source Directory Prompt
        in_folder_lbl = tk.Label(inputs_frame, text="Folder with 835s:",
                                 anchor='w')
        in_folder_lbl.grid(row=1, column=0, sticky='WE', padx=5, pady=2)

        self.in_folder_txt = tk.Entry(inputs_frame, state="disabled")
        self.in_folder_txt.grid(row=1, column=1, columnspan=7,
                                sticky="WE", padx=5, pady=2)

        self.in_folder_btn = tk.Button(inputs_frame, text="Browse ...",
                                       command=self.browse_for_open_loc)
        self.in_folder_btn.grid(row=1, column=10, sticky='E', padx=5, pady=2)

        # Save Results Prompt
        out_folder_lbl = tk.Label(inputs_frame, text="Save Results to:",
                                  anchor='w')
        out_folder_lbl.grid(row=2, column=0, sticky='WE', padx=5, pady=2)

        self.out_folder_txt = tk.Entry(inputs_frame, state="disabled")
        self.out_folder_txt.grid(row=2, column=1, columnspan=7, sticky="WE",
                                 padx=5, pady=2)

        self.out_folder_btn = tk.Button(inputs_frame, text="Browse ...",
                                        command=self.browse_for_save_loc)
        self.out_folder_btn.grid(row=2, column=10, sticky='E', padx=5, pady=2)

        # Results Output Display
        self.output_text = tk.scrolledtext.ScrolledText(
            output_frame, wrap='word', height=5, width=10,
            font=('', 8), fg="#333333")
        self.output_text.grid(row=0, column=1, sticky="NSEW", padx=5, pady=2)

        self.xscroll_bar = tk.Scrollbar(output_frame, orient='horizontal',
                                        command=self.output_text.xview)
        self.output_text.configure(xscrollcommand=self.xscroll_bar.set)
        self.xscroll_bar.grid(row=2, column=1, sticky='EW')

        # Progress Bar
        self.progress_bar = ttk.Progressbar(progress_frame, orient="horizontal",
                                            length=200, mode="determinate")

        # Run and Close       
        self.ok_btn = tk.Button(footer_frame, text="Run",
                                command=self.setup_processing)
        self.ok_btn.grid(row=2, column=2, sticky='SE', padx=5, pady=2, ipadx=27)

        close_btn = tk.Button(footer_frame, text="Close", command=self.quit)
        close_btn.grid(row=2, column=3, sticky='SE', padx=5, pady=2, ipadx=20)

    def open_changelog(self):
        """Opens changelog file for user in new window."""
        try:
            with open(self.resource_path(self.__CHANGELOG),
                      mode='r') as changelogfile:
                msg = changelogfile.read()
            new_window = tk.Toplevel(self.master)
            new_window.title('Change Log')
            new_window.resizable(width=False, height=False)
            new_window.iconbitmap(self.icon_path)

            changelog_frame = tk.Frame(new_window)
            changelog_frame.pack()

            txt_widget = tk.scrolledtext.ScrolledText(changelog_frame,
                                                      wrap='none',
                                                      state='disabled')
            txt_widget.configure(state='normal', font='TkFixedFont')
            txt_widget.insert(tk.END, str(msg))
            txt_widget.configure(state='disabled')

            xscroll_bar = tk.Scrollbar(changelog_frame, orient='horizontal',
                                       command=txt_widget.xview)
            txt_widget.configure(xscrollcommand=xscroll_bar.set)

            txt_widget.grid(row=0, column=0, sticky='EW')
            xscroll_bar.grid(row=1, column=0, sticky='EW')
        except IOError:
            tk.messagebox.showerror(title="Error",
                                    message='Error opening changelog')

    def open_help(self):
        """Opens help file for user in new window."""
        try:
            with open(self.resource_path(self.__HELPFILE), mode='r') as helpfile:
                msg = helpfile.read()
            tk.messagebox.showinfo('Help', message=msg, icon='question')
        except IOError:
            tk.messagebox.showerror(title="Error",
                                    message='Error opening help file')

    def quit(self):
        """Closes app."""
        root.destroy()

    def update_outfile_path(self, save_loc, filename):
        """Updates outfile to specified save_loc and filename."""
        self.__outfile_path = os.path.join(save_loc, filename)

    def browse_for_open_loc(self):
        """Opens window for user to navigate and select input folder location."""
        msg = "Browse for Folder containing 835s to parse..."
        open_loc = os.path.normpath(askdirectory(title=msg))
        self.in_folder_txt.configure(state='normal')
        self.in_folder_txt.delete(0, tk.END)
        self.in_folder_txt.insert(0, str(open_loc))
        self.in_folder_txt.configure(state='disabled')
        self.__source_dir = os.path.normpath(open_loc)

    def browse_for_save_loc(self):
        """Opens window for user to navigate and select output folder location."""
        save_loc = os.path.normpath(askdirectory(
            # initialdir=expanduser(pathvar.get()), # used for debugging
            title="Browse for where to save output results..."))
        self.out_folder_txt.configure(state='normal')
        self.out_folder_txt.delete(0, tk.END)
        self.out_folder_txt.insert(0, str(save_loc))
        self.out_folder_txt.configure(state='disabled')
        self.update_outfile_path(save_loc, self.__outfile_name)

    def write_to_csv(self, *args):
        """Appends input args to outfile."""
        with open(self.__outfile_path, newline='', mode='a') as outcsv:
            csv_writer = csv.writer(outcsv, delimiter=',',
                                    quotechar='"', quoting=csv.QUOTE_MINIMAL)
            data = []
            for x in args:
                data.append(x)
            csv_writer.writerow(data)

    @staticmethod
    def warn_missing_loc():
        """Shows user warning of missing input and/or output values"""
        msg = "Please specify an input and output folder location"
        tk.messagebox.showerror(title="Error- Did you forget??", message=msg)

    def print_output(self, text):
        """Adds parameter text at bottom of output text widget for user to see."""
        self.output_text.configure(state='normal')
        self.output_text.insert(tk.END, str(text))
        self.output_text.see(tk.END)
        self.output_text.configure(state='disabled')

    @staticmethod
    def resource_path(relative_path):
        # Get absolute path to resource, works for dev and for PyInstaller
        base_path = getattr(sys, '_MEIPASS',
                            os.path.dirname(os.path.abspath(__file__)))
        return os.path.join(base_path, relative_path)

    def disable_widgets(self):
        """
        Disables certain widgets from user interaction while program is parsing
        files to prevent unexpected issues."""
        self.ok_btn.configure(state='disabled')
        self.in_folder_btn.configure(state='disabled')
        self.out_folder_btn.configure(state='disabled')
        self.file_pattern_txt.configure(state='disabled')
        self.file_menu.entryconfigure(0, state='disabled')
        self.file_menu.entryconfigure(1, state='disabled')
        self.settings_menu.entryconfigure(0, state='disabled')
        self.settings_menu.entryconfigure(1, state='disabled')

    def enable_widgets(self):
        """Re-enables user interaction with certain widgets previously disabled. """
        self.ok_btn.configure(state='normal')
        self.in_folder_btn.configure(state='normal')
        self.out_folder_btn.configure(state='normal')
        self.file_pattern_txt.configure(state='normal')
        self.file_menu.entryconfigure(0, state='normal')
        self.file_menu.entryconfigure(1, state='normal')
        self.settings_menu.entryconfigure(0, state='normal')
        self.settings_menu.entryconfigure(1, state='normal')

    @staticmethod
    def check_outfile_exists(outfile_name):
        """Returns if outfile_name param is an existing file."""
        return os.path.isfile(outfile_name)

    def get_new_outfile_name(self):
        """Returns new filename affixed with current time-related elements."""
        return (self.__OUTFILE_PREFIX + " " +
                time.strftime("%Y-%m-%d-%H%M%S") + ".csv")

    def begin_progressbar(self):
        """
        Disable widgets that affect input values used when processing and
        make progress bar visible.
        """
        self.disable_widgets()
        self.progress_bar.grid(row=2, column=1, stick='EW')

    def update_progressbar(self, amount):
        """Update progress bar to specified amount value."""
        self.progress_bar['value'] = amount
        self.master.update_idletasks()

    def end_progressbar(self):
        """ Set progress bar to 0 value and hide it from user."""
        self.update_progressbar(int(0))
        self.enable_widgets()
        self.progress_bar.grid_forget()

    def process_queue(self):
        """ Unused. Potentially use when implementing a Queue. """
        pass
        # try:
        #     msg = self.queue.get(0)
        #     # Show result of the task if needed
        #     print(msg)
        #     self.progressBar.stop()
        # except queue.Empty:
        #     self.master.after(100, self.process_queue)

    def get_files_list(self, source_dir):
        """Returns a list of files in specified source_dir."""
        if self.__traverse_subdir.get():
            files = [os.path.join(root, f) for root, dirs, files
                     in os.walk(source_dir) for f in files]
            # files = [f for root,dirs,files in os.walk(source_dir) for f in files]
        else:
            files = [f for f in os.listdir(source_dir)
                     if os.path.isfile(os.path.join(source_dir, f))]
        return files

    def parse_835(self, full_file_path, filename):
        """
        Consider if filename in specified full_file_path  is split across
        multiple lines (standard) or all in 1 line (less common). Write desired
        values out as each CLP or PLB segment is found.
        Note: Currently will update TRN and N1* as the respective segment
        occurs in a file, but in the future it may be desired to clear values
        if a new one occurs, so as not to mix.
        """
        file_trn02 = file_trn03 = file_payer = file_payee = file_npi = ""
        with open(full_file_path, 'r') as file:
            file_data = file.readlines()
            num_lines_in_file = len(file_data)
            if num_lines_in_file == 1:
                file_content = file_data[0].split(sep="~")
            else:
                file_content = file_data
            for line in file_content:
                claim, clp02, plb = "", "", ""  # reset values
                if line.startswith("TRN"):
                    file_trn02 = re.sub('~', '', line.split('*')[2])  # TRN;02
                    file_trn03 = re.sub('~', '', line.split('*')[3])  # TRN;03
                if line.startswith("N1*PR"):
                    file_payer = re.sub('~', '', line.split('*')[2]).rstrip()  # N1;02
                if line.startswith("N1*PE"):
                    file_payee = re.sub('~', '', line.split('*')[2]).rstrip()  # N1;02
                    file_npi = re.sub('~', '', line.split('*')[4])  # N1;04
                if line.startswith("CLP"):
                    claim = re.sub('~', '', line.split('*')[1])  # CLP;01
                    clp02 = re.sub('~', '', line.split('*')[2])  # CLP;02
                    parsed_line = [filename, file_trn02, file_trn03,
                                   file_payer, file_payee, file_npi, claim,
                                   clp02, plb]
                    self.write_to_csv(*parsed_line)
                elif line.startswith("PLB"):  # PLB;*
                    plb = re.sub('~', '', line.rstrip())
                    parsed_line = [filename, file_trn02, file_trn03,
                                   file_payer, file_payee, file_npi, claim,
                                   clp02, plb]
                    self.write_to_csv(*parsed_line)

    def process_835s(self, file_pattern, source_dir):
        """
        Get number of files in specified source_dir and iterate through each file of
        specified file_pattern in source_dir and call function to parse the file. Increment
        files parsed and update progress bar with each 1% change in progress.
        """
        files = self.get_files_list(source_dir)
        # total_file_count = sum(len(files) for _, _, files in os.walk(source_dir)) # for debugging
        total_file_count = len(files)
        processed_file_count = 0
        for file in files:
            processed_file_count += 1
            if file.endswith(file_pattern) or (file_pattern in file):
                self.print_output(f'Reading file: {file}\n')
                full_file_path = os.path.abspath(os.path.join(source_dir, file))
                self.parse_835(full_file_path, file)
            progress = int(100 * (processed_file_count / total_file_count))
            if self.progress_bar['value'] < progress:
                self.update_progressbar(progress)
        self.end_progressbar()

    def begin_processing(self):
        """
        Creates with headers or opens existing outfile and calls function to
        process files in source dir with set file pattern. Will increment run
        counter.
        """
        with open(self.__outfile_path, newline='', mode='a') as outcsv:
            csv_writer = csv.writer(outcsv, delimiter=',', quotechar='"',
                                    quoting=csv.QUOTE_MINIMAL)
            if not self.__file_exists:
                csv_writer.writerow(self.__OUTFILE_HEADERS)
        self.print_output(f'Beginning processing...\n')
        self.print_output(f'Parsing with file pattern: {self.__file_pattern}\n')
        self.process_835s(self.__file_pattern, self.__source_dir)
        self.print_output(f'Completed stripping 835s in: {self.__source_dir}\n')
        self.print_output(f'Results saved to: {self.__outfile_path}')
        self.__run_counter += 1

    def setup_processing(self):
        """
        Checks if file pattern and in/out folder values have been specified else stops
        and warns user to specify them. If user re-running program, prompts if re-run
        is desired to avoid accidental re-runs on large amounts of files. If re-running,
        will reset output filename if setting set to NOT append results to initial run's
        file.
        """
        if len(self.file_pattern_txt.get()) > 0:
            self.__file_pattern = self.file_pattern_txt.get()
        if (len(self.in_folder_txt.get()) == 0 or
                len(self.out_folder_txt.get()) == 0):
            self.warn_missing_loc()
        else:
            ok_to_run = False
            if self.__run_counter != 1:
                if tk.messagebox.askyesno(
                        "Confirm Run", "Are you sure you want to run again?"):
                    if not self.__append_runs.get():
                        self.__outfile_name = self.get_new_outfile_name()
                        self.__file_exists = self.check_outfile_exists(
                            self.__outfile_name)
                        self.update_outfile_path(self.out_folder_txt.get(),
                                                 self.__outfile_name)
                    ok_to_run = True
                    self.print_output(f'\n\n\n*****Starting Run #{self.__run_counter} *****\n')
            else:
                ok_to_run = True
            if ok_to_run:
                self.begin_progressbar()
                # self.queue = queue.Queue()                  # Implement with Queues potentially
                # ThreadedTask(self.queue).start()            # Implement with Queues potentially
                threading.Thread(target=self.begin_processing,
                                 daemon=True).start()
                # self.master.after(100, self.process_queue)  # Implement with Queues potentially


class ThreadedTask(threading.Thread):
    """ Unused. Potentially use when implementing a Queue. """
    pass
    # def __init__(self, queue):
    #     threading.Thread.__init__(self)
    #     self.queue = queue
    # def run(self):
    #     print("running thread")
    #     #self.begin_processing()
    #     time.sleep(1)  # Simulate long running process
    #     self.queue.put("Task finished")


if __name__ == "__main__":
    root = tk.Tk()
    root.wm_title('835 File Parser')
    root.minsize(450, 265)
    root.columnconfigure(2, weight=1)
    root.rowconfigure(1, weight=1)
    myapp = App(root)
    root.mainloop()
